### ISSUE TITLE
Adding multi-window support for ipadOS with Scene Delegate

### ISSUE THREAD

**Issue by ivmirx** [MvvmCross/MvvmCross#4196]
Two years ago, Apple added [`UISceneDelegate`](https://developer.apple.com/documentation/uikit/uiscenedelegate?language=objc) (or just `SceneDelegate` in Swift) in iOS 13, along with the introduction of ipadOS. It partially replaces `UIApplicationDelegate`, and its primary purpose is enabling a multi-window environment on ipadOS, with each scene holding a reference to a `UIWindow` and managing its own lifecycle. Here's a [clear guide](https://learnappmaking.com/scene-delegate-app-delegate-xcode-11-ios-13/) comparing the old and the new approaches.

I got it working by creating a rather simple `MvxSceneDelegate` and implementing my own `IMvxIosSetup`, `IMvxApplicationDelegate`, and `IMvxIosViewPresenter` classes that do not create or access `UIWindow` (so it's possible to pass it as `null`). I also moved `RunAppStart()` body from `IMvxApplicationDelegate` to `MvxSceneDelegate`.

Now there are two problems I see. First, adding something like that to MvvmCross feels like an unnecessary breaking change because so many apps don't bother with iPad, let alone multi-window support. Maybe a plugin can be a solution?

Second, ipadOS does **not** have a concept of a focused window, period. That means that the Presenter will have no idea which window it should use for navigation after a request. This is where I got stuck and don't see any elegant solution. The view has either to send some platform-specific payload to the VM or to somehow mark the current `MvxSceneDelegate` as "last used" (subclassing `UIButton` may help).

Suggestions are welcome, if there's any way we can get it in – I will try to implement it and create a PR. If not, maybe it would be useful to write a separate docs page explaining how to hack around with subclassing and whatnot.

### Platforms affected (mark all that apply)
- [+] :iphone: iOS
- [ ] :robot: Android
- [ ] :checkered_flag: WPF
- [ ] :earth_americas: UWP
- [ ] :apple: MacOS
- [ ] :tv: tvOS
- [ ] :monkey: Xamarin.Forms
- [ ] :🌍: Blazor

**Comment by Cheesebaron** [comment]
I would definitely not mind a community contribution for this. I haven't had the need to support multiple windows on iPad yet.

As for tracking which window, ViewControllers should present. It is an interesting problem, and maybe we should from MvvmCross provide some kind of good default, but let the developer easily change the behavior.

**Comment by ivmirx** [comment]
I'm not at the point of developing an iPad version either – working with desktops for now after getting a proof of concept for each platform. But I hope to get there by the end of the year.

A good default may be the last opened window, but as soon as we get to something like this split view in Outlook, it becomes tricky to manage because two instances of the same view/viewmodel classes can be opened in two "focused" windows. Since tapping a button or a tab will stop the touch event from going up the [responder chain](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/using_responders_and_the_responder_chain_to_handle_events) to their root window, the hacky way I've come up with is subclassing those UI elements (`MvxUIButton`, etc.) and sending an `NSNotification` from them before the touch event fires the binding. The Presenter observes these notifications and updates some field like `UIWindow _lastAccessedWindow` on them. This way, when a navigation request arrives, the presenter will use the correct window.

It is relatively fragile because the developer has to remember to subclass all navigation-calling UI elements, but unless Apple suddenly adds its own "last accessed" property to windows/scenes in ipadOS 16, I just can't see how it could be possible to track without some kind of manual effort across the navigation calls. But maybe I'm missing some obvious workaround that UIKit allows… 🤔

![Outlook-iPad-SPlit-View-2](https://user-images.githubusercontent.com/10554114/121838663-5bb29600-cca6-11eb-8b62-f191e78be370.jpg)

**Comment by Cheesebaron** [comment]
Isn't it always clear, which Window/ViewController a specific button belongs to? Does it make sense, if I for instance press the + button on the right side, that a new ViewController opens in the left Window?

I mean finding out which ViewController a button belongs to is easy. I think finding out which Window a ViewController belongs to should be easy enough too.
So if we would just default to showing stuff in the Window where the "touch" happened, I think that would cover a lot of the use cases.

**Comment by Hackmodford** [comment]
Could you have multiple ViewPresenters? One for each window?

**Comment by ivmirx** [comment]
@Cheesebaron
> Isn't it always clear, which Window/ViewController a specific button belongs to?
> So if we would just default to showing stuff in the Window where the "touch" happened, I think that would cover a lot of the use cases.

Yes, my worry is how to get this information from a touch event to the Presenter without creating a mess. Obtaining the parent `UIWindow` is as easy as `View.Window` (which is available from a `UIButton` too).

> Does it make sense, if I for instance press the + button on the right side, that a new ViewController opens in the left Window?

If I understand you correctly, then no, a new ViewController almost always must be opened in the same window – be it navigation to a new view or to a modal window. The scenario where the user taps inside one scene/window and the navigation happens in another one may be possible, but I can't imagine a UX case for it.

@Hackmodford 
> Could you have multiple ViewPresenters? One for each window?

It will not solve the problem of the VM-level navigation not knowing which window to address. For example, on macOS the presenter just fallbacks to `NSApplication.SharedApplication.MainWindow` tracked by the operating system. There is no main or key window on ipadOS – all opened scenes are equally ready for input from the user.

But also currently MvvmCross for Mac (and, probably, for WPF) works well with multiple windows from a single presenter, while at the same time being one of the slimmest presenters across the platforms. So, it is possible to slightly tweak the iOS presenter to work with multiple windows, but it has to have the information on which of the windows was the last to receive input.

**Comment by ivmirx** [comment]
@Hackmodford the key window is not the same as the main window (and when it comes to macOS those can be different windows). In the case of ipadOS, the main window just does not exist because two windows in two scenes in the split view mode are both "main". One of them may become key if focused by the keyboard, but the user can easily tap a button calling to navigation in another window, so it's not possible to rely on `IsKeyWindow` for navigation either.

**Comment by follesoe** [comment]
Has anyone worked on this lately? We have some scenarios that could benefit from multiple windows on iPadOS (and visionOS). Thanks to @ivmirx for doing the initial research and explaining some of the complexities in trying to implement this within the existing MVVM Cross architecture.

**Comment by ivmirx** [comment]
Unfortunately, I had quit the project where I tried to solve this problem, so no solution on my side :(

**Comment by ivmirx** [comment]
A simple idea that I haven't been able to test: catching all touches and gestures with a hit-test of the whole UIWindow or the root view controller, and then storing in a custom presenter which window/VC was touched last.

I honestly do not remember well if it will affect the `UIResponder` chain or no, also if it will fire before or after a normal button/view touch. But it feels like a non-intrusive hack to determine the last interacted window.

**Comment by ivmirx** [comment]
[Apple is forcing scenes](https://developer.apple.com/documentation/technotes/tn3187-migrating-to-the-uikit-scene-based-life-cycle) upon everyone, AppDelegate will probably stop working in iOS 19:

> Soon, all [UIKit](https://developer.apple.com/documentation/uikit) based apps will be required to adopt the scene-based life-cycle, after which your app won’t launch if you don’t. While supporting multiple scenes is encouraged, only adoption of scene life-cycle is required.

**Comment by follesoe** [comment]
Ouch, al the sudden this became a pretty high priority issue 😬

**Comment by ivmirx** [comment]
Kind of, just moving to `SceneDelegate` for a single scene will not be hard, but the question how to detect multi-scene interactions remains.

**Comment by Cheesebaron** [comment]
So I added preliminary SceneDelegate support in: https://github.com/MvvmCross/MvvmCross/pull/4893 anything missing from this to support iOS 19?

**Comment by ivmirx** [comment]
Thanks, looks great! I guess, we'll learn the details in a month, at WWDC 2025.

**Comment by follesoe** [comment]
I have a proposed implementation in https://github.com/MvvmCross/MvvmCross/pull/5192 - please have a look @ivmirx @dsmitchell, if this would still be relevant for you guys.

Added the PR as this feature is something we miss, but should be in-reach given we now have SceneKit support.

**Comment by ivmirx** [comment]
@follesoe very cool, thanks a lot! One presenter per scene is straightforward and solves lots of potential problems. I still lack an app that can work in multi-window mode, so hopefully someone using it can chime in.
